src/main/haskell/kore/src/Kore/Step/ExpandedPattern.hs

```diff
@@ -59,16 +60,40 @@ import           Kore.Variables.Free
 to use when executing Kore. It consists of an "and" between a term, a
 predicate and a substitution
 -}
-data ExpandedPattern level variable = ExpandedPattern
-    { term         :: !(PureMLPattern level variable)
-    -- ^ Free-form pattern.
-    , predicate    :: !(Predicate level variable)
-    -- ^ pattern that only evaluates to Top or Bottom.
+
+type ExpandedPattern level variable = 
+    Predicated level variable (PureMLPattern level variable)
+
+data Predicated level variable child = Predicated
+    { term :: child
+    , predicate :: !(Predicate level variable)
     , substitution :: !(UnificationSubstitution level variable)
-    -- ^ special kind of predicate of the type
-    -- variable1 = term1 /\ variable2 = term2 /\ ...
     }
-    deriving (Eq, Show)
+    deriving(Eq, Ord, Show)
+
+instance Functor (Predicated level variable) where
+    fmap f (Predicated a p s) = Predicated (f a) p s
+
+instance 
+  ( MetaOrObject level
+  , SortedVariable variable
+  , Show (variable level)
+  , Eq (variable level)
+  , Given (SymbolOrAliasSorts level)
+  )
+  => Applicative (Predicated level variable) where
+    pure a = Predicated a makeTruePredicate []
+    a <*> b = Predicated
+        { term = f x 
+        , predicate = fst (predicate1 `makeAndPredicate` predicate2)
```

Comment:
@traiansf @u-- Doing this kind of joining instead of `mergePredicatesAndSubstitutions` is likely to be the right long-term solution, but, right now, we don't have the infrastructure to do the normalization part when it is actually needed. Our current code relies on things being normalized (not sure how much, exactly).

I don't think that `mergePredicatesAndSubstitutions` fits nicely in this instance. Should we try to add the on-demand normalization thing to this PR?

Reply:
I saw `mergePredicatesAndSubstitutions` and wasn't sure how to proceed from there. We can certainly add it using a `Given` constraint (which again will be completely nuked once we get proper AST annotations).

Or: can't we write a function `normalize`, such that `normalize (PredicateSubst p1 s1 `mappend` PredicateSubst p2 s2) = mergePredicatesAndSubstitutions [p1, p2] [s1, s2]`? Here `mappend` is the obvious monoid instance that combines the predicates with `mkAnd` and the substitutions with `++`.

Reply:
At any rate, since `ExpandedPattern` is now a type synonym and everything works as before, we can make these refactorings at our leisure.

Reply:
However, if we don't merge this now then the merge conflicts will keep piling up.

Reply:
Ok, could you add a code comment saying that people should not use this until we have the infrastructure to handle on-demand normalization?

This is not necessarily related to this PR, but we may want to split the Predicated type in two - NormalizedPredicated and NotNormalizedPredicated or something like that.